src/body.rs

```diff
@@ -73,118 +73,23 @@
+use std::io::{self, Cursor};
 use std::ops::{Deref, DerefMut};
 
 use crate::{configuration::Store, Extract, IntoResponse, Request, Response, RouteMatch};
 
-/// The raw contents of an http request or response.
-///
-/// A body is a stream of `BodyChunk`s, which are essentially `Vec<u8>` values.
-/// Both `Body` and `BodyChunk` values can be easily created from standard byte buffer types,
-/// using the `From` trait.
-#[derive(Debug)]
-pub struct Body {
-    inner: BodyInner,
-}
-
-type BodyStream = StreamObj<'static, Result<BodyChunk, Error>>;
-type Error = Box<dyn std::error::Error + Send + Sync>;
-pub struct BodyChunk(hyper::Chunk);
-
-impl BodyChunk {
-    pub fn as_bytes(&self) -> &[u8] {
-        (*self.0).as_ref()
-    }
-}
-
-impl From<Vec<u8>> for BodyChunk {
-    fn from(v: Vec<u8>) -> Self {
-        BodyChunk(v.into())
-    }
-}
-
-impl From<String> for BodyChunk {
-    fn from(v: String) -> Self {
-        BodyChunk(v.into())
-    }
-}
-
-#[derive(Debug)]
-enum BodyInner {
-    Streaming(BodyStream),
-    Fixed(Vec<u8>),
-}
-
-impl Body {
-    /// Create an empty body.
-    pub fn empty() -> Self {
-        Body {
-            inner: BodyInner::Fixed(Vec::new()),
-        }
-    }
-
-    /// Collect the full contents of the body into a vector.
-    ///
-    /// This method is asynchronous because, in general, it requires reading an async
-    /// stream of `BodyChunk` values.
-    pub async fn read_to_vec(&mut self) -> Result<Vec<u8>, Error> {
-        match &mut self.inner {
-            BodyInner::Streaming(s) => {
-                let mut bytes = Vec::new();
-                pin_mut!(s);
-                while let Some(chunk) = await!(s.next()) {
-                    // TODO: do something more robust than `unwrap`
-                    bytes.extend(chunk?.as_bytes());
-                }
-                Ok(bytes)
-            }
-            BodyInner::Fixed(v) => Ok(v.clone()),
-        }
-    }
-}
-
-impl From<Vec<u8>> for Body {
-    fn from(v: Vec<u8>) -> Self {
-        Self {
-            inner: BodyInner::Fixed(v),
-        }
-    }
-}
-
-impl From<hyper::Body> for Body {
-    fn from(body: hyper::Body) -> Body {
-        // TODO: handle chunk-level errors
-        let stream = Compat01As03::new(body).map(|c| match c {
-            Ok(chunk) => Ok(BodyChunk(chunk)),
-            Err(e) => {
-                let e: Error = Box::new(e);
-                Err(e)
-            }
-        });
-        Body {
-            inner: BodyInner::Streaming(StreamObj::new(Box::new(stream))),
-        }
-    }
-}
-
-impl From<BodyChunk> for hyper::Chunk {
-    fn from(chunk: BodyChunk) -> hyper::Chunk {
-        chunk.0
-    }
-}
-
-impl Into<hyper::Body> for Body {
-    fn into(self) -> hyper::Body {
-        match self.inner {
-            BodyInner::Fixed(v) => v.into(),
-            BodyInner::Streaming(s) => hyper::Body::wrap_stream(s.compat()),
-        }
+pub(crate) async fn body_to_vec(body: Body) -> Result<Vec<u8>, io::Error> {
```

Comment:
Why was this chosen to become a free function rather than a method on the body? I think from reading uses of this, the method seemed to be a bit nicer.

Reply:
Because the type moved out of the crate, so we can't provide it as an inherent method. A couple alternatives:

- Add it via an extension trait
- Add it to the http-service crate

I prefer the latter.

Reply:
D'oh, yes that makes perfect sense. Filed a tracking issue https://github.com/rust-net-web/http-service/issues/2.